### PR TITLE
Add abstact payload builder

### PR DESCRIPTION
**Pull request by RIg410**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -26,6 +26,11 @@ serde = {version = "1", features = ["derive"]}
 serde_json = "1.0"
 serde_with = {version = "3", features = ["base64"]}
 sha3 = "0.10"
+futures = "0.3.30"
+anyhow = "1.0.82"
+tokio = "1.29.1"
+log = "0.4.14"
+async-trait = "0.1.51"
```
Comment (i1i1) [inline Cargo.toml]: Do we need it though?

File: payload-builder/src/ledger.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+use crate::SlotArtifact;
+
+#[async_trait]
+pub trait Ledger {
+    /// Get the current slot.
+    async fn get_current_slot(&self) -> Result<Slot, Error>;
+    /// Get the slot artifact by slot id. If the slot with slot_id skipped, return next slot.
+    /// If the slot with slot_id has not been produced yet, return None.
+    async fn get_slot(&mut self, slot_id: Slot) -> Result<Option<SlotArtifact>, Error>;
```
Comment (i1i1) [inline payload-builder/src/ledger.rs]: This can be done without `async_trait` crate

File: payload-builder/Cargo.toml [hunk 3]
```diff
@@ -4,6 +4,16 @@ name = "lumio-payload-builder"
 edition.workspace = true
 
 [dependencies]
+lumio-rpc.workspace = true
+lumio-types.workspace = true
+
+anyhow.workspace = true
+borsh.workspace = true
+tokio.workspace = true
```
Comment (i1i1) [inline payload-builder/Cargo.toml:12]: I think you're missing `rt` feature because of `tokio::spawn`

File: payload-builder/src/service.rs [hunk 4]
```diff
@@ -71,6 +71,9 @@ impl<L: Ledger> PayloadService<L> {
     }
 
     fn print_statisitc(&self) {
+        if self.tasks.is_empty() {
+            return;
+        }
```
Comment (i1i1) [inline payload-builder/src/service.rs:76]: This is never the case for `op-sol` right?

File: payload-builder/src/ledger.rs (hunk 2 above)
Comment (RIg410) [reply payload-builder/src/ledger.rs]: No. solana ledger impl required async.

Comment (i1i1) [reply payload-builder/src/ledger.rs]: But you can use:
```rust
trait Ledger {
  fn async_fn(&self) -> impl Future<Output = ()>;
}
```

File: payload-builder/src/service.rs (hunk 4 above)
Comment (RIg410) [reply payload-builder/src/service.rs:76]: This is a rare case in production, but mainly for use in local work.

**Pull request merged**